src/include/ref_inline.h

```diff
@@ -0,0 +1,151 @@
+ */
+static WT_INLINE bool
+__wt_ref_is_root(WT_REF *ref)
+{
+    return (ref->home == NULL);
+}
+
+/*
+ * # The ref state API. #
+ *
+ * 5 macros are defined to manipulate the ref state. This is a highly sensitive field and protected
+ * via the double underscore keyword. The field should only be accessed via these macros.
+ *
+ * WT_REF_GET_STATE:
+ * Get the state of the ref, wraps a relaxed atomic volatile load. At the time of writing this
+ * comment this was done to enable TSan and to enable burying the field behind the
+ * aforementioned double underscore.
+ *
+ * WT_REF_SET_STATE:
+ * Set the ref state. If HAVE_REF_TRACK is defined, track where the set call originated from. The
+ * ref state tracking is why we use macros here, since the tracking utilizes gcc identifiers to get
+ * the function and line number where the macro was called.
+ *
+ * WT_REF_CAS_STATE:
+ * Swap in a new state to the ref, tracking where the call originated from.
+ *
+ * WT_REF_LOCK:
+ * Spin until the state WT_REF_LOCKED is swapped into the ref state field. Once the call to this
+ * function completes the caller has exclusive access to the ref.
+ *
+ * WT_REF_UNLOCK:
+ * Effectively wraps WT_REF_SET_STATE, however should only be used when returning the ref to the
+ * previous state as returned by WT_REF_LOCK.
+ */
+
+/*
+ * __ref_set_state --
+ *     Set a ref's state. Accessed from the WT_REF_SET_STATE macro.
+ */
+static WT_INLINE void
+__ref_set_state(WT_REF *ref, uint8_t state)
+{
+    WT_RELEASE_WRITE_WITH_BARRIER(ref->__state, state);
+}
+
+#ifndef HAVE_REF_TRACK
+#define WT_REF_SET_STATE(ref, s) __ref_set_state((ref), (s))
+#else
+/*
+ * __ref_track_state --
+ *     Save tracking data when REF_TRACK is enabled. This is diagnostic code and ref->state changes
+ *     are a hot path. As such we allow some racing in the history tracking code instead of
+ *     requiring a lock and slowing down ref state transitions.
+ */
+static WT_INLINE void
+__ref_track_state(
+  WT_SESSION_IMPL *session, WT_REF *ref, uint8_t new_state, const char *func, int line)
+{
+    ref->hist[ref->histoff].session = session;
+    ref->hist[ref->histoff].name = session->name;
+    __wt_seconds32(session, &ref->hist[ref->histoff].time_sec);
+    ref->hist[ref->histoff].func = func;
+    ref->hist[ref->histoff].line = (uint16_t)line;
+    ref->hist[ref->histoff].state = (uint16_t)(new_state);
+    ref->histoff = (ref->histoff + 1) % WT_ELEMENTS(ref->hist);
+}
+
+#define WT_REF_SET_STATE(ref, s)                                           \
+    do {                                                                   \
+        __ref_track_state(session, ref, s, __PRETTY_FUNCTION__, __LINE__); \
+        __ref_set_state((ref), (s));                                       \
+    } while (0)
+#endif
+
+/*
+ * __ref_get_state --
+ *     Get a ref's state variable safely.
+ */
+static WT_INLINE uint8_t
+__ref_get_state(WT_REF *ref)
+{
+    return (__wt_atomic_loadv8(&ref->__state));
```

Comment:
This is now always sequential consistency? It is not required to be that strict in most places.

Reply:
It's relaxed